src/bun.js/child_process.exports.js

```diff
@@ -1,3 +1,1622 @@
+  validateTimeout(options.timeout);
+
+  // Validate maxBuffer, if present.
+  validateMaxBuffer(options.maxBuffer);
+
+  options.killSignal = sanitizeKillSignal(options.killSignal);
+
+  const child = spawn(file, args, {
+    cwd: options.cwd,
+    env: options.env,
+    // gid: options.gid,
+    shell: options.shell,
+    signal: options.signal,
+    // uid: options.uid,
+  });
+
+  let encoding;
+  const _stdout = [];
+  const _stderr = [];
+  if (options.encoding !== "buffer" && BufferIsEncoding(options.encoding)) {
+    encoding = options.encoding;
+  } else {
+    encoding = null;
+  }
+  let stdoutLen = 0;
+  let stderrLen = 0;
+  let killed = false;
+  let exited = false;
+  let timeoutId;
+
+  let ex = null;
+
+  let cmd = file;
+
+  function exitHandler(code, signal) {
+    if (exited) return;
+    exited = true;
+
+    if (timeoutId) {
+      clearTimeout(timeoutId);
+      timeoutId = null;
+    }
+
+    if (!callback) return;
+
+    // merge chunks
+    let stdout;
+    let stderr;
+    if (encoding || (child.stdout && child.stdout.readableEncoding)) {
+      stdout = ArrayPrototypeJoin.call(_stdout, "");
+    } else {
+      stdout = BufferConcat(_stdout);
+    }
+    if (encoding || (child.stderr && child.stderr.readableEncoding)) {
+      stderr = ArrayPrototypeJoin.call(_stderr, "");
+    } else {
+      stderr = BufferConcat(_stderr);
+    }
+
+    // TODO: Make this check code === 0 when Bun.spawn fixes exit code issue
+    if (!ex && code >= 0 && signal === null) {
+      callback(null, stdout, stderr);
+      return;
+    }
+
+    if (args?.length) cmd += ` ${ArrayPrototypeJoin.call(args, " ")}`;
+
+    if (!ex) {
+      ex = genericNodeError(`Command failed: ${cmd}\n${stderr}`, {
+        // code: code < 0 ? getSystemErrorName(code) : code, // TODO: Add getSystemErrorName
+        code: code,
+        killed: child.killed || killed,
+        signal: signal,
+      });
+    }
+
+    ex.cmd = cmd;
+    callback(ex, stdout, stderr);
+  }
+
+  function errorHandler(e) {
+    ex = e;
+
+    if (child.stdout) child.stdout.destroy();
+    if (child.stderr) child.stderr.destroy();
+
+    exitHandler();
+  }
+
+  function kill() {
+    if (child.stdout) child.stdout.destroy();
+    if (child.stderr) child.stderr.destroy();
+
+    killed = true;
+    try {
+      child.kill(options.killSignal);
+    } catch (e) {
+      ex = e;
+      exitHandler();
+    }
+  }
+
+  if (options.timeout > 0) {
+    timeoutId = setTimeout(function delayedKill() {
+      kill();
+      timeoutId = null;
+    }, options.timeout);
+  }
+
+  if (child.stdout) {
+    if (encoding) child.stdout.setEncoding(encoding);
+
+    child.stdout.on("data", function onChildStdout(chunk) {
+      // Do not need to count the length
+      if (options.maxBuffer === Infinity) {
+        ArrayPrototypePush.call(_stdout, chunk);
+        return;
+      }
+      const encoding = child.stdout.readableEncoding;
+      const length = encoding
```

Comment:
imo this function should be split into two versions:
1) with an encoding set
2) no encoding set

That should make it more monomorphic  and `slice`  can be inlined + use `apply` instead of creating a temporary array and then copying the temporary array again